src/uthash.h

```diff
@@ -323,13 +375,24 @@ do {
     (add)->hh.prev = NULL;                                                       \
     (head) = (add);                                                              \
     HASH_MAKE_TABLE(hh, head);                                                   \
+    if (HASH_NOMEM_OK && !(head)->hh.tbl) {                                      \
+      (head) = 0;                                                                \
+      SET_MEM_FAILED(&(add)->hh, 1);                                             \
+      uthash_mem_failed(add);                                                    \
```

Comment:
FWIW, I still haven't figured out the logic by which sometimes we do `mem_ok = 0`, sometimes we do `SET_MEM_FAILED`, sometimes we do `uthash_mem_failed`, and sometimes we do `uthash_fatal("out of memory")`. I suspect we might be able to just get rid of `uthash_mem_failed`; but are the other 2 or 3 mechanisms really irreducible?